### PR TITLE
Compute the expected body length to reduce heap allocations

### PR DESCRIPTION
**Pull request by ebfull**
## Motivation

Unnecessary allocation(s) in the external protocol codec implementation.

## Solution

Let's precompute the expected body length of a network message and reserve this capacity in the provided buffer. Also compute the checksum in-place.

## Review

This review is a low priority.

## Follow Up

Add a size estimator #1774 
Proptest our message implementation #27
Add benchmarks for network messages #1775

### CODE REVIEW

File: zebra-network/src/protocol/external/codec.rs
Comment (teor2345) [inline zebra-network/src/protocol/external/codec.rs:181]:
```suggestion
    ///
    /// TODO: Replace with a size estimate, to avoid multiple serializations
    /// for large data structures like lists, blocks, and transactions.
    /// See #1774.
    fn body_length(&self, msg: &Message) -> usize {
```

**Review by teor2345 — DISMISSED** [review]
Thanks for this PR!

This change removes a Vec allocation for every peer message. But the exact body length calculation adds an extra serialization. (I'm not sure if the functions called by our serializations do any extra allocations internally. And I don't know how much of the redundant FakeWriter serializations will be removed by the optimiser. But our serialization code does try to write directly to the Writer.)

Overall, allocations are more expensive than serialization, particularly if the allocation incurs a page fault. So this should provide a small performance improvement. (And a bit of code cleanup.)

We should follow up this PR with a size estimator and a benchmark, but that's on us :-)

**Comment by teor2345** [comment]
The macOS testnet CI hung without any logs, #1222 should fix that, and #1140 will get us better diagnostics when that happens.

I thought we'd found all the places where Zebra just hung without any output, but maybe we've missed one.

**Comment by ebfull** [comment]
@teor2345 Yeah, I was originally writing out size estimators for each message, but eventually due to all the nested structures it started to look like... the serialization code. So I figured this would be as efficient modulo some pointer arithmetic (which probably won't get optimized out, but meh)

**Comment by teor2345** [comment]
@ebfull yeah it's possible there's no good design for size estimation - or we'd have to rewrite all out serialisation code using a visitor pattern. So we might never end up doing that ticket.

Doubling the size of all our serialisation code wouldn't be great for performance either.

**Review by teor2345 — APPROVED** [review]
The latest change just applies my comment suggestion.

**Comment by teor2345** [comment]
We should re-run CI after #1789 merges, then merge this PR once it passes.

**Pull request merged**
